src/mlpack/core/optimizers/fw/atoms.hpp

```diff
@@ -0,0 +1,202 @@
+/**
+ * @file atoms.hpp
+ * @author Chenzhe Diao
+ *
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_FW_ATOMS_HPP
+#define MLPACK_CORE_OPTIMIZERS_FW_ATOMS_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/optimizers/proximal/proximal.hpp>
+#include "func_sq.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Class to hold the information and operations of current atoms in the
+ * soluton space.
+ */
+class Atoms
+{
+ public:
+  Atoms(){ /* Nothing to do. */ }
+
+  /**
+   * Add atom into the solution space.
+   *
+   * @param v new atom to be added.
+   * @param c coefficient of the new atom.
+   */
+  void AddAtom(const arma::vec& v, const double c = 0)
+  {
+    if (currentAtoms.is_empty())
+    {
+      CurrentAtoms() = v;
+      CurrentCoeffs().set_size(1);
+      CurrentCoeffs().fill(c);
+    }
+    else
+    {
+      currentAtoms.insert_cols(0, v);
+      arma::vec cVec(1);
+      cVec(0) = c;
+      currentCoeffs.insert_rows(0, cVec);
+    }
+  }
+
+
+  //! Recover the solution coordinate from the coefficients of current atoms.
+  void RecoverVector(arma::mat& x)
+  {
+    x = currentAtoms * currentCoeffs;
+  }
+
+  /** 
+   * Prune the support, delete previous atoms if they don't contribute much.
+   * See Algorithm 2 of paper:
+   * @code
+   * @article{RaoShaWri:2015Forward--backward,
+   *    Author = {Rao, Nikhil and Shah, Parikshit and Wright, Stephen},
+   *    Journal = {IEEE Transactions on Signal Processing},
+   *    Number = {21},
+   *    Pages = {5798--5811},
+   *    Publisher = {IEEE},
+   *    Title = {Forward--backward greedy algorithms for atomic norm regularization},
+   *    Volume = {63},
+   *    Year = {2015}
+   * }
+   * @endcode
+   *
+   * @param F thresholding number.
+   * @param function function to be optimized.
+   */
+  void PruneSupport(const double F, FuncSq& function)
+  {
+    arma::mat atomSqTerm = function.MatrixA() * currentAtoms;
+    atomSqTerm = sum(square(atomSqTerm), 0);
```

Comment:
I wonder if it would be helpful and slightly more efficient to inline this into one line:

```
arma::mat atomSqTerm = arma::sum(arma::square(function.MatrixA() * currentAtoms), 0);
```

Reply:
This atomSqTerm could be calculated when each atom is added, so we don't need to recalculate all of them each time we prune the support. I just modified the implementation. Thanks for pointing out the problem.